### ISSUE TITLE
Wish we could disable eager evaluation in 1.3

### ISSUE THREAD

**Issue by ffxsam** [meteor/meteor#5988]
**_2 Upvotes**_ I really like the idea of modules and enforcing a strict load order. But since eager evaluation is still enabled, I find myself creating `imports` folders everywhere to force lazy evaluation instead. So I've got `/client/imports/components`, `/lib/imports/collections/stuff.js`, etc.

Is there some way to disable eager evaluation in a project?

**Comment by trusktr** [comment]
I believe if you ever import a file, it should not be eagerly evaluated. So, if a file is outside an `imports` folder, it should be eagerly evaluated only if it is not ever explicitly imported.

**Comment by ffxsam** [comment]
Oh, if that's the case, then that's great. Because actually I realized I do want some things to be global, such as Mongo collections. If they're defined using `export`, then if I run `meteor shell`, I'm unable to access them.

**Comment by zimt28** [comment]
Haven't used 1.3 yet, but isn't it possible to put everything into a root `/imports/[...]` folder?
@trusktr I don't think so, in that case we wouldn't need `imports` folders.
@benjamn What about a `.meteor/entry-points` config file? If it exists, lazy loading would be the default and Meteor would use the client & server files specified in that config file as entry points.

**Comment by stubailo** [comment]
I just put everything into a top-level `imports` folder and then the actual entry points in `main/client` and `main/server`.

See the app I'm working on here: https://github.com/meteor/todos

Eager evaluation was accidentally left on for code in `imports` in an old beta. In `beta.4`, no files in `imports` should be eagerly evaluated.

**Comment by ffxsam** [comment]
Oh! I'm using beta.4 actually. So this is a non-issue? If I `import` something from a file, that file will not be eagerly evaluated?

**Comment by zimme** [comment]
From what I understand, any file in an `import` folder won't be eagerly loaded, any file not in a `import` folder will.

**Comment by zimme** [comment]
Though it would be awesome to have an option to have the opposite behaviour.

Nothing is eagerly loaded, unless it's files in a `load`/`autoload` folder.

**Comment by benjamn** [comment]
Though it's tempting, the reason I have avoided treating `import`ed modules as lazy is that they suddenly become eager when/if code elsewhere in the app stops importing them. If you have a lazy module that at some point ceases to be imported by any other modules, it would be very weird for it to be start being evaluated immediately on startup. Instead, it should stop being evaluated at all.

Another option is to consider a module lazy if it `export`s anything, but again it's weird that the module suddenly becomes eager if you remove the last `export` statement. A little less weird, perhaps, because the module itself gets to decide whether it exports anything, but still potentially jarring.

**Comment by zimme** [comment]
Yeah, those two scenarios where a file is evaluated differently based on content doesn't seem intuitive.

I feel that special folders or file names is more in line with previous versions of meteor and what more meteor developers are used to.

**Comment by ffxsam** [comment]
>Another option is to consider a module lazy if it exports anything, but again it's weird that the module suddenly becomes eager if you remove the last export statement. A little less weird, perhaps, because the module itself gets to decide whether it exports anything, but still potentially jarring.

I like this idea the best. That way I can still have globals, for things like collections that I'd want to access via `meteor shell`. But if I have a JS file with `export`s in it, then Meteor doesn't evaluate it at all, not until another file `import`s it.

**Comment by stubailo** [comment]
I think it's totally reasonable to have lazy-loaded modules that only exist for side effects, which would be impossible if having `export` in a file made it lazy.

**Comment by ffxsam** [comment]
On a semi-related note, would it be possible to modify the Meteor shell so that we could modularlize things like collections, but still use them in the shell? e.g.

```
> import Posts from '/lib/collections/posts';
> Posts.remove({});
```

**Comment by benjamn** [comment]
@ffxsam the shell transpiles commands if `ecmascript` is installed (see [here](https://github.com/meteor/meteor/blob/42a6bad2e3942ff44fa99e6108d930e8db628a5c/tools/static-assets/server/shell-server.js#L326-L350)), and `require` is now global (after d542ffb1a96418b190d84999c5802a157dacea66), so I *think* what you're hoping for *might* already work?

**Comment by ffxsam** [comment]
I don't have time at the moment to double check, but I basically cloned git@github.com:mantrajs/mantra-sample-blog-app.git and ran it, and was not able to use `import` nor `require` to access `Posts` in the shell.

**Comment by trusktr** [comment]
@stubailo Where's the entry points in that todo app? Is any JavaScript inside the `main` folder treated as an entry point?

**Comment by stubailo** [comment]
Yes, any JS file outside of an `imports/` directory is an entry point. So I decided to name that folder `main` for clarity.

**Comment by rozzzly** [comment]
I'd love the option to specify just an one entry point for client, and one for server, then have everything else be lazy. That's what I've between doing with `universe: modules-entrypoint` and its between awesome. 

Another possibility which I've suggested before was a `.gitignore`-esque setup for specifying which files should be loaded lazily/eagerly. Although I'm not sure that I'd be into that myself...

------

As far as possibly letting the user configure easy/lazy loading. I suppose there might be some loooongterm things to consider. So the `package.js`'s `api` let's us specify a package's main module... Could we do something similar for the app root? A central place to store config/build logic for the app as a whole?

There's not really an `app.js`… would that be something to explore? Or is there a reason why that doesn't exist? Lol I suppose that's all going to eventuality be part of bundling, wherever that gets addressed.

**Comment by trusktr** [comment]
> I'd love the option to specify just an one entry point for client, and one for server, then have everything else be lazy. That's what I've between doing with universe:modules-entrypoint and its between awesome.

You can already do that. Just place everything to be "lazy-loaded" into an `imports` folder, and only the two client/server entry points outside of it.

**Comment by jamiter** [comment]
Not sure if this is the right place, but someone suggested a `autoimport` packages that will have the Meteor 1.2 behaviour when it comes to eager loading. Removing this package would make everything lazy. What do you think about this?

**Comment by zimme** [comment]
@jamiter: That would be nice, but what would be the entry points when `autoimport` is removed?

**Comment by jamiter** [comment]
That could be import.js, as it is now. It could be index.js like in npm packages. Or we could put it in the package.json under the main attribute. Just naming some things. Any other ideas are welcome :)

**Comment by clayne11** [comment]
+1. I like this idea.

**Comment by rozzzly** [comment]
+1 @jamiter :+1: 

This is how it's done in `universe:modules`; there's an additional (optional) package `universe:modules-entrypoint` which forces the loader into a single entry point from which you use can do `System.import(...)` conditionally--which is important for server/client code--to direct loading and evaluation. 

That way, _picky purists_ such as myself can have a nice clean and dogmatic file structure. And average developers, maintainers of existing codebases, etc, etc... can migrate gradually. 

@trusktr I know that would be possible with the current proposal. But, that directory structure... just ew. Just as bad as forcing a extension suffix like **`MyFileName.import.js`**.

**Comment by ffxsam** [comment]
I've had some time to think on this a bit.. I've slightly changed my stance.

I personally would like for certain objects such as `Meteor` to always be global. It's kind of a pain to have to import it everywhere, and I think Meteor devs (especially programmers new to JS) are used to be able to just use it.

**Comment by trusktr** [comment]
@rozzzly Just don't create any `imports` folders or leave unused files laying around, then it's pure. ;)

**Comment by jamiter** [comment]
@ffxsam, you can make anything global by doing this:
```javascript
import thing from 'thing'

global.thing = thing;
```
This could be part of your `main.js` or any other file that is loaded first. This way you decide yourself if variables need to be global or not. And personally I think it's better to have a file where you can see what is explicitly set global instead of thing (randomly) being global.

It might be a preference to have things global or not, but it's still possible with the new modules. Making everything import/export only by default won't stop you from having your preferred globals.

**Comment by ffxsam** [comment]
@jamiter Ah, nice.. thanks for the tip! Though just to point out, `main.js` is loaded last.

**Comment by rozzzly** [comment]
Sorry @truskt, I don't follow when you say:
> Just don't create any imports folders or leave unused files laying around, then it's pure. ;)

Are you suggesting the absence of an imports folder disables eager-loading? I don't think that's what you're saying. As stubailo said,
> Yes, any JS file oue of an imports/ directory is an entry point. So I decided to name that folder main for clarity.

lol rephrase plz

**Comment by trusktr** [comment]
@rozzzly What I mean is that if you don't have an `imports` folder, and all your files are imported into some entry point (f.e. `client/main.js` or `server/main.js`), then all of your files will be "lazy-loaded", and only your main.js files will be eagerly evaluated. If you have any other files that are not ever imported, those are eagerly loaded too (they simply execute in a script tag in the head). But, the definition that Meteor 1.3 is using for " lazy" loading I think may be confusing, because I think of "lazy" loading as not needing to evaluate a script until it is needed, for example when a route changes (using f.e. SystemJS' `System.import` to load the script right when needed). Meteor 1.3, if I'm not mistaken "lazy" loads a script if you have `import`ed or `require`ed it, but, nonetheless the script will still be evaluated as soon as the client loads your main.js, so it's not really "lazy" in the sense that I take "lazy" to be. Basically, in Meteor 1.3, eager means auto-loaded as with pre 1.3, and lazy means you define the load order based on where you import things (not really lazy in the sense that I consider to be true "lazy" loading).

**Comment by hwillson** [comment]
This has become a long thread with several different ideas/suggestions in it, but the original issue has been addressed by Meteor 1.3 (see https://github.com/meteor/meteor/issues/5988#issuecomment-171420274 for an explanation of why this wasn't working properly). As for the many other ideas/suggestions in this thread, if anyone feels strongly about them and would like to see those ideas live on, please consider opening a new issue to track them. As things stand however, the original issue has been addressed (and there has been no activity on this issue thread in over a year), so I'll close it off. Thanks for posting this issue originally and the great discussion all!
